From d7d1eebd4d5806be55ffacbf18917ad68d4ae7fd Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Tue, 31 May 2022 12:51:43 -0300
Subject: [PATCH 126/127] iconv: Use 64 bit stat for gconv_parseconfdir (BZ#
 29213)

The issue is only when used within libc.so (iconvconfig already builds
with _TIME_SIZE=64).

This is a missing spot initially from 52a5fe70a2c77935.

Checked on i686-linux-gnu.

(cherry picked from commit c789e6e40974e2b67bd33a17f29b20dce6ae8822)
---
 NEWS                       | 1 +
 iconv/gconv_parseconfdir.h | 9 ++++++---
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/NEWS b/NEWS
index beaaf7a6c2..8cea3b52d3 100644
--- a/NEWS
+++ b/NEWS
@@ -39,6 +39,7 @@ The following bugs are resolved with this release:
   [29209] libc: isfdtype is not y2038 aware
   [29210] network: ruserpass is not y2038 aware
   [29211] libc: __open_catalog is not y2038 aware
+  [29213] libc: gconv_parseconfdir is not y2038 aware
 
 
 Version 2.35
diff --git a/iconv/gconv_parseconfdir.h b/iconv/gconv_parseconfdir.h
index c0de548833..debb96b322 100644
--- a/iconv/gconv_parseconfdir.h
+++ b/iconv/gconv_parseconfdir.h
@@ -32,8 +32,11 @@
 # define readdir __readdir
 # define closedir __closedir
 # define mempcpy __mempcpy
-# define lstat64 __lstat64
+# define struct_stat struct __stat64_t64
+# define lstat __lstat64_time64
 # define feof_unlocked __feof_unlocked
+#else
+# define struct_stat struct stat
 #endif
 
 /* Name of the file containing the module information in the directories
@@ -158,12 +161,12 @@ gconv_parseconfdir (const char *prefix, const char *dir, size_t dir_len)
 	      && strcmp (ent->d_name + len - strlen (suffix), suffix) == 0)
 	    {
 	      char *conf;
-	      struct stat64 st;
+	      struct_stat st;
 	      if (asprintf (&conf, "%s/%s", buf, ent->d_name) < 0)
 		continue;
 
 	      if (ent->d_type != DT_UNKNOWN
-		  || (lstat64 (conf, &st) != -1 && S_ISREG (st.st_mode)))
+		  || (lstat (conf, &st) != -1 && S_ISREG (st.st_mode)))
 		found |= read_conf_file (conf, dir, dir_len);
 
 	      free (conf);
-- 
2.36.1

